### ISSUE TITLE
Benchmark single-file read from a large zipped folder

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#1711]
New requirement for #1702 

---

Let's create the following structure

```
maindir/
> file_1.json   # (size: FILESIZE characters)
> file_2.json
> file_3.json
> subfolder1/
> > file_1.json
> > ...
> > file_IMAGES.json
> subfolder2/ # identical to subfolder1
> subfolder3/ # identical to subfolder1
> subfolder4/ # identical to subfolder1
> subfolder5/ # identical to subfolder1
```

Then we have two parameters, FILESIZE and IMAGES.

Tests should run for 
1. FILESIZE equivalent to 10k, IMAGES=100
2. FILESIZE equivalent to 500k, IMAGES=100
3. FILESIZE equivalent to 10k, IMAGES=10000

(sanity check: we should not look at a zipped folder taking up more than 250M)

---

Test: access `subfolder2/file_1.json` and time this operation.

**Comment by tcompa** [comment]
Additional information: what is the compression factor for the three cases?

**Comment by ychiucco** [comment]
Benchmark script:
```python
import os
import random
import shutil
import string
from devtools import debug
from time import perf_counter
from pathlib import Path
from zipfile import ZipFile
from zipfile import ZIP_DEFLATED

def generate_file(size_in_kilobytes: int, folder: Path):
    name = ''.join(
        random.choices(
            string.ascii_lowercase + string.ascii_uppercase + string.digits,
            k=15,
        )
    )
    with open(folder / name, 'w') as file:
        file.write("_"*1_024*size_in_kilobytes)

def setup_repo(file_size_in_kilobytes: int, number_of_images: int) -> Path:
    
    root = Path("zip_benchmark")
    root.mkdir()

    for _ in range(3):
        generate_file(file_size_in_kilobytes, root)
    
    for i in range(5):
        subfolder = root / f"subfolder{i+1}"
        subfolder.mkdir()
        for _ in range(number_of_images):
            generate_file(file_size_in_kilobytes, subfolder)

    return root

def zip_repo(root: Path) -> Path:
    zip_file = Path(f"{root}.zip")
    with ZipFile(
        zip_file, mode="w", compression=ZIP_DEFLATED
    ) as zipfile:
        for root_dir, _, files in os.walk(root):
            for file in files:
                file_path = os.path.join(root_dir, file)
                archive_path = os.path.relpath(file_path, root_dir)
                zipfile.write(file_path, archive_path)
    return zip_file

def time_to_read_single_file_from_zip(zip_path: str) -> float:
    total_time = 0

    with ZipFile(zip_path, 'r') as zip_file:
        for file_name in zip_file.namelist():

            start_reading = perf_counter()
            with zip_file.open(file_name) as file:
                file.read().decode('utf-8')
            end_reading = perf_counter()
            total_time += end_reading - start_reading

    return total_time / len(zip_file.namelist())

def run_benchmark(
    file_sizes: list[int], numbers_of_images: list[int]
) -> dict[str, float]:

    results = dict()

    for file_size in file_sizes:
        for number_of_images in numbers_of_images:

            root = setup_repo(file_size, number_of_images)

            start_zipping = perf_counter()
            zip_file = zip_repo(root)
            end_zipping = perf_counter()
            shutil.rmtree(root)

            results[(file_size, number_of_images)] = dict(
                zipping_time = end_zipping - start_zipping,
                read_time = time_to_read_single_file_from_zip(zip_file)
            )

            zip_file.unlink()

    return results

if __name__ == "__main__":
    results = run_benchmark(
        file_sizes=[10, 100, 1_000],
        numbers_of_images=[10, 100, 1_000],
    )
    debug(results)
```

Results:
```console
results: {
        (10, 10): {
            'zipping_time': 0.012625302999367705,
            'read_time': 3.2992245391107884e-05,
        },
        (10, 100): {
            'zipping_time': 0.09711969699856127,
            'read_time': 3.397895425964734e-05,
        },
        (10, 1000): {
            'zipping_time': 1.0037251579997246,
            'read_time': 3.265411395660158e-05,
        },
        (100, 10): {
            'zipping_time': 0.02756120099729742,
            'read_time': 0.00011355932066129204,
        },
        (100, 100): {
            'zipping_time': 0.24275659799968707,
            'read_time': 8.58504771212487e-05,
        },
        (
            100,
            1000,
        ): {
            'zipping_time': 2.460784101000172,
            'read_time': 8.350354389759964e-05,
        },
        (1000, 10): {
            'zipping_time': 0.20028994500171393,
            'read_time': 0.0005259049435410833,
        },
        (
            1000,
            100,
        ): {
            'zipping_time': 1.8383918079998693,
            'read_time': 0.0004924211291295767,
        },
        (
            1000,
            1000,
        ): {
            'zipping_time': 18.500729942999897,
            'read_time': 0.0005718442474808132,
        },
    } (dict) len=9
```

**Comment by ychiucco** [comment]
Benchmark script:
```python
import os
import random
import shlex
import shutil
import string
import subprocess
from pathlib import Path
from time import perf_counter
from zipfile import ZipFile

from fractal_server.zip_tools import _zip_folder_to_file_and_remove


def generate_file(ind_folder, ind_name, size_in_kilobytes: int, folder: Path):
    with open(folder / f"{ind_folder}_{ind_name}", "w") as file:
        file.write(
            "".join(
                random.choices(
                    (
                        string.ascii_lowercase
                        + string.ascii_uppercase
                        + string.digits
                    ),
                    k=1_024 * size_in_kilobytes,
                )
            )
        )


def generate_folder(
    file_size_in_kilobytes: int, number_of_images: int
) -> Path:
    root = Path("root")
    root.mkdir()

    for ind_file in range(3):
        generate_file("_", ind_file, file_size_in_kilobytes, root)

    for ind_folder in range(5):
        subfolder = root / f"subfolder{ind_folder+1}"
        subfolder.mkdir()
        for ind_file in range(number_of_images):
            generate_file(
                ind_folder, ind_file, file_size_in_kilobytes, subfolder
            )

    return root


def time_to_read_single_file_from_zip(zip_path: str) -> float:
    iterations = 10
    total_time = 0
    filename = "subfolder3/2_3"
    start_reading = perf_counter()
    for it in range(iterations):
        with ZipFile(zip_path, "r") as zip_file:
            with zip_file.open(filename) as file:
                file.read().decode("utf-8")
    end_reading = perf_counter()
    total_time += end_reading - start_reading
    return total_time / iterations


def run_benchmark(
    file_sizes_in_kb: list[int], numbers_of_images: list[int]
) -> dict[str, float]:

    print(
        "File size\tNum images\tFolder size\tZipfile size\tCompression\tZipping time\tSingle-file-access time"
    )
    for file_size_in_kb in file_sizes_in_kb:
        for number_of_images in numbers_of_images:

            folder = generate_folder(file_size_in_kb, number_of_images)
            res = subprocess.run(
                shlex.split(f"du -s {folder}"),
                capture_output=True,
                encoding="utf8",
            )
            folder_size_kbytes = int(res.stdout.split("\t")[0])
            folder_size_Mbytes = folder_size_kbytes / 1_024

            zip_file = f"{folder}.zip"
            start_zipping = perf_counter()
            _zip_folder_to_file_and_remove(folder=folder)
            end_zipping = perf_counter()
            zipping_time = end_zipping - start_zipping
            zipfile_size_bytes = os.path.getsize(zip_file)
            zipfile_size_Mbytes = zipfile_size_bytes / 1_000_000
            compression_ratio = zipfile_size_Mbytes / folder_size_Mbytes

            single_file_read_time = time_to_read_single_file_from_zip(zip_file)

            print(f"{file_size_in_kb:6d} kB", end="\t")
            print(f"{number_of_images:5d}", end="\t")
            print(f"{folder_size_Mbytes:8.3f} M", end="\t")
            print(f"{zipfile_size_Mbytes:6.4f} M", end="\t")
            print(f"{compression_ratio:5.4f}", end="\t")
            print(f"{zipping_time:9.4f} s", end="\t")
            print(f"{single_file_read_time:17.4f} s", end="\n")

            Path(zip_file).unlink()


if __name__ == "__main__":
    results = run_benchmark(
        file_sizes_in_kb=[10, 100, 1000],
        numbers_of_images=[10, 100, 1000],
    )
```

Results:
```console
File size  Num images  Folder size  Zipfile size  Compression  Zipping time  Single-file-access time
    10 kB          10      1.242 M      0.4137 M       0.3330      0.0465 s                 0.0004 s
    10 kB         100     11.789 M      3.9271 M       0.3331      0.2885 s                 0.0026 s
    10 kB        1000    117.258 M      39.0705 M       0.3332      2.9371 s                 0.0248 s
   100 kB          10     10.352 M      4.0878 M       0.3949      0.2234 s                 0.0008 s
   100 kB         100     98.242 M      38.7962 M       0.3949      2.2284 s                 0.0029 s
   100 kB        1000    977.148 M      385.8900 M       0.3949     20.7799 s                 0.0259 s
  1000 kB          10    103.516 M      40.8205 M       0.3943      2.2142 s                 0.0031 s
  1000 kB         100    982.422 M      387.4108 M       0.3943     22.4195 s                 0.0058 s
  1000 kB         1000     ¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?¿?
```

**Comment by ychiucco** [comment]
```
File size  Num images  Folder size  Zipfile size  Compression  Zipping time  Single-file-access time
    10 kB          10      1.242 M      0.2810 M       0.2262      0.0535 s                 0.0008 s
    10 kB         100     11.789 M      2.6669 M       0.2262      0.3780 s                 0.0025 s
    10 kB         200     23.508 M      5.3201 M       0.2263      0.7138 s                 0.0059 s
    10 kB         300     35.227 M      7.9714 M       0.2263      1.0015 s                 0.0073 s
   100 kB          10     10.352 M      2.6197 M       0.2531      0.3680 s                 0.0006 s
   100 kB         100     98.242 M      24.8623 M       0.2531      3.3509 s                 0.0026 s
   100 kB         200    195.898 M      49.5769 M       0.2531      6.6823 s                 0.0046 s
   100 kB         300    293.555 M      74.2937 M       0.2531     11.5280 s                 0.0075 s
   200 kB          10     20.703 M      5.1668 M       0.2496      0.8806 s                 0.0009 s
   200 kB         100    196.484 M      49.0367 M       0.2496      7.2328 s                 0.0027 s
   200 kB         200    391.797 M      97.7811 M       0.2496     14.7046 s                 0.0053 s
   200 kB         300    587.109 M      146.5287 M       0.2496     22.9184 s                 0.0074 s
   300 kB          10     31.055 M      7.7124 M       0.2483      1.2331 s                 0.0012 s
   300 kB         100    294.727 M      73.1970 M       0.2484     11.6824 s                 0.0035 s
   300 kB         200    587.695 M      145.9578 M       0.2484     22.5843 s                 0.0051 s
   300 kB         300    880.664 M      218.7203 M       0.2484     34.8582 s                 0.0076 s
```

**Comment by tcompa** [comment]
The last comment by @ychiucco  has some answer to the question about how long it takes to read a single file from a zipped job folder.

# Scaling

The single-file access time (last column in tables below) scales linearly with the number of files (which in the table is named "Num images", although this is actually the number of files, and a single image would correspond to a handful of files) and it seems to not scale at all with file size:

```
File size  Num images  Folder size  Zipfile size  Compression  Zipping time  Single-file-access time
    10 kB          10      1.242 M      0.2808 M       0.2260      0.0555 s                 0.0008 s
    10 kB         100     11.789 M      2.6673 M       0.2262      0.3448 s                 0.0031 s
    10 kB        1000    117.258 M      26.5363 M      0.2263      3.5140 s                 0.0275 s
    10 kB       10000   1171.945 M      265.3404 M     0.2264     34.9033 s                 0.3067 s
```

```
File size  Num images  Folder size  Zipfile size  Compression  Zipping time  Single-file-access time
    10 kB         300     35.227 M      7.9714 M       0.2263      1.0015 s                 0.0073 s
   100 kB         300    293.555 M     74.2937 M       0.2531     11.5280 s                 0.0075 s
   200 kB         300    587.109 M    146.5287 M       0.2496     22.9184 s                 0.0074 s
   300 kB         300    880.664 M    218.7203 M       0.2484     34.8582 s                 0.0076 s
```

# Absolute times

The largest reported example is a zip with 50 thousand 10kB files (it's 10 thousand files per folder, with 5 folders):
```
File size  Num images  Folder size  Zipfile size  Compression  Zipping time  Single-file-access time
    10 kB       10000   1171.945 M      265.3404 M     0.2264     34.9033 s                 0.3067 s
```
Each image-list items corresponds at worst (no SLURM-job batching at all) to five files (args.json, logs, slurm stdout/stderr, sbatch script). This means that 50 thousands files represent a single task running on 10 thousand images, or e.g. 5 tasks running on 2 thousand images. In this situation, reading a single file from the zip takes quite long (~300 ms).

If we have a job with 10 thousand images and 10 tasks, then we are quickly getting into 10 seconds response times (this is a worst-case scaling scenario, with no job batching at all).

Is this acceptable? It depends - of course - on where this access operation would be performed (that is, would the corresponding endpoint require very short response times?).

Note: the reported times only matter for when a job is over. When the job is running, the non-compressed folder still exists and there is no problem in accessing its files. Does this mitigate the risk of long responses?


# TLDR

The fact that these times are under control for all typical use cases and the fact that we don't currently have any feature performing this operation make me think that we should proceed anyway with #1702 - which has the immediate benefit of reducing our footprint (in terms of disk usage and file numbers). We can then keep in mind that developing a feature requiring fast single-file access would require also some improvements here.

However, if this seems like a blocking point for important future feature, then we should simply not merge #1702. We can clearly keep a copy of its core changes available for the future, which may be useful in cleaning up this feature even in its current form.

Feedback is welcome - cc @jluethi

**Comment by tcompa** [comment]
> For the record, it looks (similar to what we observed in other contexts) that calling unzip directly takes quite shorter than using Python zipfile library to extract a single file from a zipped folder with many files. Of course this can be explored/used as a workaround, if the above-mentioned performance is not acceptable.

We confirmed this guess with @ychiucco. Here are some simple data, where the main message is quite clear:
```
number_of_images=2000
Python: 0.0377 s
Unzip:  0.0016 s

number_of_images=4000
Python: 0.0845 s
Unzip:  0.0022 s

number_of_images=6000
Python: 0.1722 s
Unzip:  0.0035 s

number_of_images=8000
Python: 0.2065 s
Unzip:  0.0038 s

number_of_images=10000
Python: 0.2704 s
Unzip:  0.0041 s
```

and here is the corresponding script
```python
import os
import random
import shlex
import shutil
import string
import subprocess
from pathlib import Path
from time import perf_counter
from zipfile import ZipFile, ZipInfo

from fractal_server.zip_tools import _zip_folder_to_file_and_remove

random.seed(123123213)

def generate_file(ind_folder, ind_name, size_in_kilobytes: int, folder: Path):
    with open(folder / f"{ind_folder}_{ind_name}", "w") as file:
        file.write(
            # f"{ind_name}" * int(1_024 * size_in_kilobytes))

            "".join(
                random.choices(
                    (
                        string.ascii_lowercase
                        + string.ascii_uppercase
                        + string.digits
                    ),
                    k=int(1_024 * size_in_kilobytes),
                )
            )
        )


def generate_folder(
    file_size_in_kilobytes: int, number_of_images: int
) -> Path:
    root = Path("root")
    root.mkdir()

    for ind_file in range(3):
        generate_file("_", ind_file, file_size_in_kilobytes, root)

    for ind_folder in range(5):
        subfolder = root / f"subfolder{ind_folder+1}"
        subfolder.mkdir()
        for ind_file in range(number_of_images):
            generate_file(
                ind_folder, ind_file, file_size_in_kilobytes, subfolder
            )

    return root

INTERNAL_FILENAME = "subfolder3/2_3"



def time_to_read_single_file_from_zip_python(zip_path: str) -> float:
    iterations = 10
    total_time = 0
    start_reading = perf_counter()
    for it in range(iterations):
        with ZipFile(zip_path, "r") as zip_file:
            with zip_file.open(INTERNAL_FILENAME) as file:
                output = file.read().decode("utf-8")
    end_reading = perf_counter()
    total_time += end_reading - start_reading
    return total_time / iterations, output


def time_to_read_single_file_from_zip_unzip(zip_path: str) -> float:
    iterations = 10
    total_time = 0
    start_reading = perf_counter()
    for _ in range(iterations):
        res = subprocess.run(
            shlex.split(f"unzip -p {zip_path} {INTERNAL_FILENAME}"),
            capture_output=True,
            encoding="utf-8",
        )
        res.stdout
    end_reading = perf_counter()
    total_time = end_reading - start_reading
    return total_time / iterations, res.stdout


def run_benchmark():
    
    file_size_in_kb = 0.1
    for number_of_images in [2_000, 4_000, 6_000, 8_000, 10_000]:
        print(f"{number_of_images=}")

        folder = generate_folder(file_size_in_kb, number_of_images)

        zip_file = f"{folder}.zip"
        _zip_folder_to_file_and_remove(folder=folder)

        single_file_read_time_python, out_py = time_to_read_single_file_from_zip_python(zip_file)
        single_file_read_time_unzip, out_unzip = time_to_read_single_file_from_zip_unzip(zip_file)

        assert out_py == out_unzip

        print(f"Python: {single_file_read_time_python:.4f} s", end="\n")
        print(f"Unzip:  {single_file_read_time_unzip:.4f} s", end="\n")
        
        Path(zip_file).unlink()
        try:
            shutil.rmtree(folder)
        except:
            pass
        print()

run_benchmark()
```

**Comment by tcompa** [comment]
Given the workaround described in the last comment, performance for single-file access is not a blocking issue any more (for future features) - thus we'll merge #1702.
